parser/context.go

```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2017 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parser
+
+import (
+	"fmt"
+
+	"github.com/google/go-jsonnet/ast"
+)
+
+var topLevelContext = "$"
+
+const anonymous = "anonymous"
+
+// TODO(sbarzowski) polish children functions and consider moving to AST
+// and exporting
+
+// directChildren are children of AST node that are executed in the same context
+// and environment as their parent
+func directChildren(node ast.Node) []ast.Node {
+	switch node := node.(type) {
+	case *ast.Apply:
+		return []ast.Node{node.Target}
+	case *ast.ApplyBrace:
+		return []ast.Node{node.Left, node.Right}
+	case *ast.Array:
+		return nil
+	case *ast.Assert:
+		return []ast.Node{node.Cond, node.Message, node.Rest}
+	case *ast.Binary:
+		return nil
```

Comment:
I assume you're talking about boolean short cuts.  They're not thunks because they are executed immediately (if they are executed at all).  The connection with thunks is that in a lazy language you can implement short cuts with a function, but that doesn't mean they are thunks when implemented as a construct instead of a function.